resolver/caching_resolver.go

```diff
@@ -111,21 +111,33 @@ func (r *CachingResolver) reloadCacheEntry(ctx context.Context, cacheKey string)
 	logger.Debugf("prefetching '%s' (%s)", util.Obfuscate(domainName), qType)
 
 	req := newRequest(dns.Fqdn(domainName), qType)
+
 	response, err := r.next.Resolve(ctx, req)
+	if err != nil {
+		util.LogOnError(ctx, fmt.Sprintf("can't prefetch '%s' ", domainName), err)
 
-	if err == nil {
-		if response.Res.Rcode == dns.RcodeSuccess {
-			packed, err := response.Res.Pack()
-			if err != nil {
-				logger.Error("unable to pack response", err)
+		return nil, 0
+	}
 
-				return nil, 0
-			}
+	if response.Res.Rcode == dns.RcodeSuccess {
+		respCopy := response.Res.Copy()
 
-			return &packed, r.adjustTTLs(response.Res.Answer)
+		// don't cache any EDNS OPT records
+		util.RemoveEdns0Record(respCopy)
+
+		packed, err := respCopy.Pack()
+		if err != nil {
+			logger.Error("unable to pack response", err)
+
+			return nil, 0
 		}
-	} else {
-		util.LogOnError(ctx, fmt.Sprintf("can't prefetch '%s' ", domainName), err)
+
+		if r.redisClient != nil {
+			res := *respCopy
+			r.redisClient.PublishCache(cacheKey, &res)
+		}
+
+		return &packed, r.adjustTTLs(response.Res.Answer)
```

Comment:
This is also a slight difference compared to the `putInCache` flow: here Redis sees the unadjusted TTLs whereas when we add values using `putInCache` Redis gets the adjusted TTLs.  
My initial reaction is this is correct, and the other code should change since other blocky instances might have different TTL configuration.

Reply:
Sending the unadjusted TTL to Redis isn't a problem since received entries are TTL adjusted upon receiving before they are put in the local cache. 

This way instances with different configs can  share the same sync channel without breaking each other's settings. 

That was the base idea when I implemented it. I get that this is counterintuitive and should be changed.